### PR TITLE
Fix the showBottomSheet controller leaking

### LINKED ISSUE #87708: `Scaffold.of(context).showBottomSheet<T>` leak the internal `AnimationController`
https://github.com/flutter/flutter/blob/575043fc8bfda5bc6b20444885cb84a6acb563e0/packages/flutter/lib/src/material/scaffold.dart#L2882-L2895

This internal `AnimationController` will be a leak because no one handles its life circle when calling `showBottomSheet` multe times.

### Repro Code
```dart
 testWidgets('Each of the internal animation controller should be disposed by the framework.', (WidgetTester tester) async {
    final GlobalKey<ScaffoldState> scaffoldKey = GlobalKey();
    await tester.pumpWidget(MaterialApp(
      home: Scaffold(
        key: scaffoldKey,
        body: const Center(child: Text('body')),
      ),
    ));

    scaffoldKey.currentState!.showBottomSheet<void>((_) {
      return Builder(
        builder: (BuildContext context) {
          return Container(height: 200.0);
        },
      );
    });

    await tester.pump();
    expect(find.byType(BottomSheet), findsOneWidget);

    // The first sheet's animation is still running.

    // Trigger the second sheet will remove the first sheet from tree.
    scaffoldKey.currentState!.showBottomSheet<void>((_) {
      return Builder(
        builder: (BuildContext context) {
          return Container(height: 200.0);
        },
      );
    });
    await tester.pump();
    expect(find.byType(BottomSheet), findsOneWidget);

    // Remove the Scaffold from the tree.
    await tester.pumpWidget(const SizedBox.shrink());

    // If the internal animation controller do not dispose will throw
    // FlutterError:<ScaffoldState#1981a(tickers: tracking 1 ticker) was disposed with an active
    // Ticker.
    expect(tester.takeException(), isNull);
  });
```

### PR DESCRIPTION
**Pull request by xu-baolin**
Fixes #87708

### CODE REVIEW

**Comment by xu-baolin** [comment]
@Piinks hi, could you help review this? Thanks so much.

File: packages/flutter/test/material/bottom_sheet_test.dart [hunk 1]
```diff
@@ -832,6 +832,105 @@ void main() {
...
+    // Remove the Scaffold from the tree.
+    await tester.pumpWidget(const SizedBox.shrink());
+
+    // If the internal animation controller do not dispose will throw
+    // FlutterError:<ScaffoldState#1981a(tickers: tracking 1 ticker) was disposed with an active
+    // Ticker.
+    expect(tester.takeException(), isNull);
+  });
+
+  testWidgets('the framework do not dispose the transitionAnimationController provided by user.', (WidgetTester tester) async {
```
Comment (Piinks) [inline packages/flutter/test/material/bottom_sheet_test.dart:878]: Nit:
```suggestion
  testWidgets('The framework does not dispose of the transitionAnimationController provided by user.', (WidgetTester tester) async {
```

File: packages/flutter/test/material/bottom_sheet_test.dart [hunk 2]
```diff
@@ -832,6 +832,105 @@ void main() {
     expect(find.text('BottomSheet'), findsNothing);
   });
 
+  // Regression test for https://github.com/flutter/flutter/issues/87708
```
Comment (Piinks) [inline packages/flutter/test/material/bottom_sheet_test.dart:835]: Nit: Can you add this comment to the other test as well?

File: packages/flutter/test/material/bottom_sheet_test.dart [hunk 3]
```diff
@@ -832,6 +832,105 @@ void main() {
...
+  testWidgets('Each of the internal animation controller should be disposed by the framework.', (WidgetTester tester) async {
```
Comment (Piinks) [inline packages/flutter/test/material/bottom_sheet_test.dart:836]:
```suggestion
  testWidgets('Each of the internal animation controllers should be disposed by the framework.', (WidgetTester tester) async {
```

File: packages/flutter/lib/src/material/scaffold.dart [hunk 4]
```diff
@@ -2690,6 +2690,7 @@ class ScaffoldState extends State<Scaffold> with TickerProviderStateMixin, Resto
     ShapeBorder? shape,
     Clip? clipBehavior,
     BoxConstraints? constraints,
+    bool willDisposeAnimationController = true,
```
Comment (Piinks) [inline packages/flutter/lib/src/material/scaffold.dart:2693]: Maybe a bit more readable
```suggestion
    bool shouldDisposeAnimationController = true,
```

**Review by Piinks — COMMENTED** [review]
Thanks for this @xu-baolin! Just a few nits, great fix!

File: packages/flutter/test/material/bottom_sheet_test.dart (hunk 2 above)
Comment (xu-baolin) [reply packages/flutter/test/material/bottom_sheet_test.dart:898]: done

File: packages/flutter/test/material/bottom_sheet_test.dart (hunk 1 above)
Comment (xu-baolin) [reply packages/flutter/test/material/bottom_sheet_test.dart]: done

File: packages/flutter/lib/src/material/scaffold.dart (hunk 4 above)
Comment (xu-baolin) [reply packages/flutter/lib/src/material/scaffold.dart]: done

**Review by Piinks — APPROVED** [review]
LGTM! Thank you!

**Pull request merged**